### DISCUSSION TITLE
Tools and governance.

### DISCUSSION THREAD

**Discussion by gabriel-lucas** [Open-Telecoms-Data/open-fibre-data-standard#289] [Ideas]
Hi everyone,

We are in the process of organizing our network and it would be very valuable to have some website / tool to create and edit our network information following the OFDS schema (nodes, spans, etc).
It would be valuable also to have the ERD diagram and sql for creating a database and populating it with our data. Is there something similar?  I have seen the data format conversion tool, but is not what I´m looking for.

How many organizations are publishing their networks using OFDS? 
Is there an open list or database?

Lastly, I think it would be very valuable to look at what has been done in the open data transportation / mobility space.

In 2006 the GTFS [1] format was created and it has been adopted in more than 70 countries and there more than 2000 data feeds [2] had been published. The Mobility Data non profit was created and they take care of the specification, tools, etc.

Links:
[1] - https://gtfs.org/
[2] - https://mobilitydatabase.org/
[3] - https://mobilitydata.org/history/
[4] - https://mobilitydata.org/what-we-do/

**Comment by stevesong** [comment]
Hi Gabriel,

Thanks for writing! 
 
> We are in the process of organizing our network and it would be very valuable to have some website / tool to create and edit our network information following the OFDS schema (nodes, spans, etc). It would be valuable also to have the ERD diagram and sql for creating a database and populating it with our data. Is there something similar? I have seen the data format conversion tool, but is not what I´m looking for.

An ERD and sql statement for creating a database is a great idea!  We don't have anything like that currently but I think it would very obviously create value.  Is this something you might be willing to work on?  I would be happily to help in any way I can.

> How many organizations are publishing their networks using OFDS? Is there an open list or database?

Currently there are no organisations 'officially' using OFDS although many are testing it out internally.  There are some [sample datasets](https://github.com/stevesong/OFDS-datasets) available but these have not been validated by operators.   

> Lastly, I think it would be very valuable to look at what has been done in the open data transportation / mobility space.
> 
> In 2006 the GTFS [1] format was created and it has been adopted in more than 70 countries and there more than 2000 data feeds [2] had been published. The Mobility Data non profit was created and they take care of the specification, tools, etc.
> 
> Links: [1] - https://gtfs.org/ [2] - https://mobilitydatabase.org/ [3] - https://mobilitydata.org/history/ [4] - https://mobilitydata.org/what-we-do/

Yes! This is exactly the sort of mechanism and governance model that I think would work well for OFDS.  I mention GTFS directly in my most [recently talk](https://www.afpif.org/wp-content/uploads/2024/08/Introducing-the-Open-Fibre-Standard.pdf) on OFDS. I'd be very interested in talking to your further about it.

**Comment by duncandewhurst** [comment]
Hi Gabriel,

Good questions :-)

Currently, no OFDS-specific tool exists for creating and editing network information according to the schema. However, OFDS offers a [GeoJSON publication format](https://open-fibre-data-standard.readthedocs.io/en/latest/reference/publication_formats/geojson.html) and import, export and editing of GeoJSON data is widely supported by GIS tools like QGIS and ArcGIS. We experimented with creating a [QGIS template](https://github.com/Open-Telecoms-Data/ofds_qgis_template) ([draft documentation](https://docs.google.com/document/d/18Hmh60shJiM6CMo7VadvzK_KBb6kgB3NN_BBuNKntoE/edit?tab=t.0#heading=h.32u932t73olg)) for creating data based on full OFDS schema, but we ran into various issues with the QGIS forms interface which made it difficult to use. You might be able to create something simpler and more usable if you are only interested in a subset of OFDS fields.

We don't have an ERD or SQL statements to create a database right now, but the [CSV publication format](https://open-fibre-data-standard.readthedocs.io/en/latest/reference/publication_formats/csv.html) could be used as a blueprint for a relational database. You could use the example CSV files provided in the documentation in combination with a tool like csvkit's [csvsql](https://csvkit.readthedocs.io/en/latest/scripts/csvsql.html) command to create SQL tables based on the files. However, you would still need to create the foreign key relationships based on the relationships listed in the documentation. Alternatively, you could look at something like [jsonschema2db](https://github.com/better/jsonschema2db) to create a database schema directly from the OFDS JSON Schema, although I don't know how well it works.

A word of caution:  like GTFS, OFDS is designed as a format for publishing and sharing data. It is not designed for 'internal' data storage in an operational system. There are likely to be other things to consider when designing a database for a system that manages information about fibre networks.

If you choose to pursue designing a database around the OFDS schema, I think the first step should be identifying which OFDS fields and relationships are relevant to you. OFDS has many fields, some of which may be irrelevant, and it is designed to accommodate different [network business models](https://open-fibre-data-standard.readthedocs.io/en/latest/primer/scopeandkeyconcepts.html#network-business-models), which leads to some complexity (e.g. one-to-many relationships) that you might be able to simplify in your own database.

**Comment by gabriel-lucas** [comment]
Hi Steve, Dunca,

Thank you very much for your very informative responses and for the links, I have not had the time to study them deeply but I managed to create the [ERD / SQL for creating the database.](https://github.com/gabriel-lucas/OFDS/blob/main/ofds.sql). 
I did it by hand, because I wanted to understand better every table and field.
I also managed to create a first prototype of an  [online editor](https://ofds.retool.com/apps/b73bef36-9948-11ef-aac5-2fdc81a865d4/OFDS/networks), using retool.com.

It would be great to have a good set of tools, such as online editor, API´s, public databases, that can integrate with the pipelines of actors in this field (isp´s, regulators, etc)

Take a look to this:
- [Internet Census 2012](http://internetcensus2012.github.io/InternetCensus2012/paper.html)
- and this library [leafmap](https://leafmap.org).

**Comment by ramon-roca** [comment]
Hello everyone, I heard about this thread while talking to Gabriel, and glad to see some old friends here ;)

As some of you may already know, the fact is that we are working on the implementation of systems to support community and open networks, convinced that these tools are key to being able to go beyond the classic/restricted proprietary/commercial implementation point of view.

You can get an idea of ​​the initiative by reading the introduction that is currently available at https://network.guifi.net

Obviously such an implementation is based on a data model, in this case it started about 20 years ago somewhat successfully for wireless networks, and during the period 2020-2022 there was a review for the optical fiber we already had a quite few years implementing but without the support of equivalent tools, once we realized the impact caused by not having it.

Next week I'm moving to Timor-Leste and I'll have to put a lot of focus on the implementation there, but beyond that I'm realizing that maybe it's time to review some aspects of the data model towards standardization, and on this occasion take not only the needs that we have been detecting, also get inspired by the OFDS and try, whenever possible, converge or contribute.

These days I have been looking at it. A first thing that I see is that by having to put the focus on an implementation, this forces to define a more extensive scheme to support the operational needs that probably goes beyond the scope that the OFDS initially intended.

Within the framework of this exercise, I already observe many similarities in terms of the basic scheme, but also observing those aspects that go further and that I would like to share and happy to hear thoughts/comments.

As a first example:

-Investors, owners... Consider that these roles can be extended to individual people, not just organizations, and organizations can also be made up of people (which can move from organizations)

-The implementation of the operation of an open network with optical fiber also implies new roles beyond owners or network operators, like service providers, which are the services (Internet, VoIP, TV...) and their supported profiles (bandwidth, ...). Some operation roles might be delegated (subcontracted)  for different criteria, for example maintenance in geographical areas or capabilities.

Well, maybe I'm making a too long post and not sure if it's in the right context, but in any case I want to take the opportunity to say hello and share some thoughts that were coming to my mind ;)

**Reply by stevesong** [reply]
Hola Ramon. I would love to see the standard accommodate Guifi.net's requirements. 

Regarding including individuals not just organisations, I think we could include individuals as "minimum viable organisations" :-)  for the time being but look at something more inclusive in the next iteration of the standard.

Regarding the inclusion of digital service providers (such as VoIP, streaming, IP, etc) we have focused on the physical infrastructure of the fibre networks so far (and their ownership, investment) for OFDS but complementary data structures that profile data service providers sounds great.  For me there is still a lot to get right about the physical layer yet though.  Once the foundation is solid, this seems like a natural extension.

I think we need some basic tools to begin with such as what Gabriel is [working on](https://github.com/gabriel-lucas/OFDS).
